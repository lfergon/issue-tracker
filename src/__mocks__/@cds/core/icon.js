module.exports = {
  ClarityIcons: {
    addIcons: jest.fn()
  }
};