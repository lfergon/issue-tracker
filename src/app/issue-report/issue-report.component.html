<h3>
  Report an issue
</h3>
<form clrForm *ngIf="issueForm" [formGroup]="issueForm" (ngSubmit)="addIssue()">
  <clr-input-container>
    <label>Issue</label>
    <input clrInput formControlName="title">
    <clr-control-error>Title is required</clr-control-error>
  </clr-input-container>
  <div class="clr-row" *ngIf="suggestions.length">
    <div class="clr-col-lg-2"></div>
    <div class="clr-col-lg-6">
      <clr-stack-view>
        <clr-stack-header>Similar issues</clr-stack-header>
        <clr-stack-block *ngFor="let suggestion of suggestions">
          <clr-stack-label>{{suggestion.issueNumber}} : {{suggestion.title}}</clr-stack-label>
          <clr-stack-content>
            <p>{{suggestion.description}}</p>
          </clr-stack-content>
        </clr-stack-block>
      </clr-stack-view>
    </div>
  </div>
  <clr-textarea-container>
    <label>Description</label>
    <textarea clrTextarea formControlName="description" placeholder="Description"></textarea>
  </clr-textarea-container>
  <clr-radio-container clrInline>
    <label>Priority</label>
    <clr-radio-wrapper>
      <input clrRadio type="radio" name="priority" value="low" formControlName="priority">
      <label>Low</label>
    </clr-radio-wrapper>
    <clr-radio-wrapper>
      <input clrRadio type="radio" name="priority" value="medium" formControlName="priority">
      <label>Medium</label>
    </clr-radio-wrapper>
    <clr-radio-wrapper>
      <input clrRadio type="radio" name="priority" value="high" formControlName="priority">
      <label>High</label>
    </clr-radio-wrapper>
    <clr-control-error>Priority is required</clr-control-error>
  </clr-radio-container>
  <clr-select-container>
    <label>Type</label>
    <select clrSelect formControlName="type">
      <option value="feature">Feature</option>
      <option value="bug">Bug</option>
      <option value="task">Task</option>
    </select>
    <clr-control-error>Type is required</clr-control-error>
  </clr-select-container>
  <button class="btn btn-primary" type="submit">Create</button>
</form>
