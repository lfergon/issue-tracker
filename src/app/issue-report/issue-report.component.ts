import { Component, OnInit, inject, output } from '@angular/core';
import {
  FormControl,
  FormGroup,
  Validators,
  ReactiveFormsModule,
} from '@angular/forms';
import { IssuesService } from '../issues.service';
import { Issue } from '../issue';
import {
  ClrInputModule,
  ClrRadioModule,
  ClrSelectModule,
  ClrStackViewModule,
  ClrTextareaModule,
} from '@clr/angular';
import { CommonModule } from '@angular/common';

interface IssueForm {
  title: FormControl<string>;
  description: FormControl<string>;
  priority: FormControl<string>;
  type: FormControl<string>;
}

@Component({
  selector: 'app-issue-report',
  templateUrl: './issue-report.component.html',
  styleUrls: ['./issue-report.component.css'],
  imports: [
    ClrInputModule,
    ReactiveFormsModule,
    ClrStackViewModule,
    ClrTextareaModule,
    ClrRadioModule,
    ClrSelectModule,
    CommonModule,
  ],
})
export class IssueReportComponent implements OnInit {
  readonly formClose = output();

  suggestions: Issue[] = [];

  issueForm = new FormGroup<IssueForm>({
    title: new FormControl('', {
      nonNullable: true,
      validators: Validators.required,
    }),
    description: new FormControl('', { nonNullable: true }),
    priority: new FormControl('', {
      nonNullable: true,
      validators: Validators.required,
    }),
    type: new FormControl('', {
      nonNullable: true,
      validators: Validators.required,
    }),
  });

  private issuesService = inject(IssuesService);

  addIssue(): void {
    if (this.issueForm && this.issueForm.invalid) {
      this.issueForm.markAllAsTouched();
      return;
    }
    this.issuesService.createIssue(this.issueForm.getRawValue() as Issue);
    this.formClose.emit();
  }

  ngOnInit(): void {
    this.issueForm.controls.title.valueChanges.subscribe((title) => {
      this.suggestions = this.issuesService.getSuggestions(title);
    });
  }
}
