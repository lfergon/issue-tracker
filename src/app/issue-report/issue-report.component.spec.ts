import { ComponentFixture, TestBed } from '@angular/core/testing';

import { IssueReportComponent } from './issue-report.component';
import { IssuesService } from '../issues.service';
import { ReactiveFormsModule } from '@angular/forms';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';

describe('IssueReportComponent', () => {
  let component: IssueReportComponent;
  let fixture: ComponentFixture<IssueReportComponent>;
  let mockIssuesService: { createIssue: jest.Mock; getSuggestions: jest.Mock };

  beforeEach(async () => {
    mockIssuesService = {
      createIssue: jest.fn(),
      getSuggestions: jest.fn(),
    };

    await TestBed.configureTestingModule({
      declarations: [IssueReportComponent],
      imports: [ReactiveFormsModule],
      providers: [{ provide: IssuesService, useValue: mockIssuesService }],
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
    }).compileComponents();

    fixture = TestBed.createComponent(IssueReportComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should emit formClose event when addIssue is called and form is valid', () => {
    component.issueForm.controls.title.setValue('Test title');
    component.issueForm.controls.description.setValue('Test description');
    component.issueForm.controls.priority.setValue('High');
    component.issueForm.controls.type.setValue('Bug');

    const emitSpy = jest.spyOn(component.formClose, 'emit');

    component.addIssue();
    expect(emitSpy).toHaveBeenCalled();
  });

  it('should get suggestions based on title value changes', () => {
    const mockSuggestions = [
      { title: 'Test', description: '', priority: '', type: '' },
    ];
    mockIssuesService.getSuggestions.mockReturnValue(mockSuggestions);
    component.issueForm.controls.title.setValue('Test');
    expect(component.suggestions).toEqual(mockSuggestions);
  });
});
