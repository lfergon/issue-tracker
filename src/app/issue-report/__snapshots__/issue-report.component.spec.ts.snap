// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssueReportComponent should match snapshot 1`] = `
<app-issue-report
  formClose={[Function EventEmitter_]}
  issueForm={[Function FormGroup]}
  issuesService={[Function Object]}
  suggestions={[Function Array]}
>
  <h3>
    Report an issue
  </h3><form
    class="ng-untouched ng-pristine ng-invalid"
    clrform=""
    novalidate=""
  >
    <clr-input-container>
      <label>
        Issue
      </label>
      <input
        class="ng-untouched ng-pristine ng-invalid"
        clrinput=""
        formcontrolname="title"
      />
      <clr-control-error>
        Title is required
      </clr-control-error>
    </clr-input-container>
    <clr-textarea-container>
      <label>
        Description
      </label>
      <textarea
        class="ng-untouched ng-pristine ng-valid"
        clrtextarea=""
        formcontrolname="description"
        placeholder="Description"
      />
    </clr-textarea-container>
    <clr-radio-container
      clrinline=""
    >
      <label>
        Priority
      </label>
      <clr-radio-wrapper>
        <input
          class="ng-untouched ng-pristine ng-invalid"
          clrradio=""
          formcontrolname="priority"
          name="priority"
          type="radio"
          value="low"
        />
        <label>
          Low
        </label>
      </clr-radio-wrapper>
      <clr-radio-wrapper>
        <input
          class="ng-untouched ng-pristine ng-invalid"
          clrradio=""
          formcontrolname="priority"
          name="priority"
          type="radio"
          value="medium"
        />
        <label>
          Medium
        </label>
      </clr-radio-wrapper>
      <clr-radio-wrapper>
        <input
          class="ng-untouched ng-pristine ng-invalid"
          clrradio=""
          formcontrolname="priority"
          name="priority"
          type="radio"
          value="high"
        />
        <label>
          High
        </label>
      </clr-radio-wrapper>
      <clr-control-error>
        Priority is required
      </clr-control-error>
    </clr-radio-container>
    <clr-select-container>
      <label>
        Type
      </label>
      <select
        class="ng-untouched ng-pristine ng-invalid"
        clrselect=""
        formcontrolname="type"
      >
        <option
          value="feature"
        >
          Feature
        </option>
        <option
          value="bug"
        >
          Bug
        </option>
        <option
          value="task"
        >
          Task
        </option>
      </select>
      <clr-control-error>
        Type is required
      </clr-control-error>
    </clr-select-container>
    <button
      class="btn btn-primary"
      type="submit"
    >
      Create
    </button>
  </form>
</app-issue-report>
`;
